swarm/src/lib.rs

```diff
@@ -422,6 +425,7 @@ where TBehaviour: NetworkBehaviour<ProtocolsHandler = THandler>,
     /// Returns `true` if the address existed and was removed, `false`
     /// otherwise.
     pub fn remove_external_address(&mut self, addr: &Multiaddr) -> bool {
+        self.behaviour.inject_expired_external_addr(addr);
         self.external_addrs.remove(addr)
```

Comment:
```suggestion
        if self.external_addrs.remove(addr) {
            self.behaviour.inject_expired_external_addr(addr);
        }
```
However, note also that the most common way in which external addresses expire is as a result of new addresses being added. See https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/registry.rs#L186-L196. So, to be able to properly invoke this callback whenever an external address expires, I think you would need to incorporate the expired/removed addresses within a call to `add()` in the `AddAddressResult`, so that the `Swarm` can then invoke the callback.